woodpile/src/pile_reader.rs

```diff
@@ -0,0 +1,1617 @@
+    }
+
+    #[inline(always)]
+    fn make_empty(pile_dir: PathBuf, max_record_size: usize) -> PileReader {
+        PileReader::new_from_source_files(pile_dir, vec![], Some(max_record_size))
+    }
+
+    #[inline(never)]
+    fn open_impl(
+        log_directory: PathBuf,
+        time: time::PrimitiveDateTime,
+        now: VouchedTime,
+        options: PileReaderOptions,
+        start_offsets: Option<&HashMap<PathBuf, u64>>,
+        wait_for_summary: bool,
+    ) -> Result<PileReaderState> {
+        let empty = HashMap::new();
+        let start_offsets = start_offsets.unwrap_or(&empty);
+
+        let (pile_dir, write_deadline) = crate::construct_epoch_subdirectory(log_directory, time);
+
+        let close_time = write_deadline
+            .saturating_add(crate::EPOCH_WRITE_LEEWAY)
+            .saturating_add(crate::CLOCK_ERROR_BOUND);
+        let past_closing = now.get_local_time() > close_time;
+
+        // Past closing time, we might have a summary.
+        if past_closing {
+            use crate::close::SnapshotFileStatus;
+
+            // Generating the path for the summary file also flushes the NFS caches if
+            // necessary (just enough so we don't get spurious ENOENT on the hierarchy).
+            let (summary_path, exists) = crate::close::closed_subdir_summary_path(pile_dir.clone());
+
+            let mut have_summary_file = exists == SnapshotFileStatus::Present;
+
+            // Handle a missing directory like an empty pile.
+            if !have_summary_file
+                && matches!(pile_dir.metadata(), Err(e) if e.kind() == std::io::ErrorKind::NotFound)
+            {
+                // The missing time bucket directory is old enough
+                // that it should stay missing.
+                let inner = PileReader::make_empty(pile_dir, options.max_record_size);
+                return Ok(PileReaderState::Stable(StablePileReader { inner }));
+            }
+
+            let force_close_time = close_time
+                .saturating_add(options.force_close_grace_period.max(time::Duration::ZERO));
+            if (!have_summary_file) & (now.get_local_time() > force_close_time) {
+                // Must forcibly close the directory;
+
+                // Can't do that in read-only mode.
+                if options.mode == PileReaderMode::ReadOnly {
+                    return Err(std::io::Error::other(
+                        "Pile directory past closing time, but PileReader is read-only.",
+                    ));
+                }
+
+                let fsync = options.mode == PileReaderMode::Fsync;
+                let _ = crate::close::close_epoch_subdir(
+                    pile_dir.clone(),
+                    now,
+                    crate::close::CloseEpochOptions { fsync },
+                )?;
+
+                have_summary_file = true;
+            }
+
+            if have_summary_file {
+                let (inner, rewound) = Self::new_from_zip_archive(
```

Comment:
i don't think we have enough information here: we know where to resume reading, we need to know where the last read happened.

Reply:
actually, the code is fine. we read each file by stat-ing it for the length, and then read everything up to that length. This means we can catch at most in-flight write (we assume each log shard file is single writer, otherwise NFS is lossy) at the end, so we only need to remember the last correctly decoded record.

There could be a hole where we find the sentinel but raced past the earlier payload writes, for the record. shard_reader should address that, when the last record is invalid but right at the end of the region.

Reply:
(it would also be a broken filesystem that lets you see the size update before the data update... it's just gonna show you garbage? on NFS, getting the updated stat results should also flush any cached data)

Reply:
I think we can still remember the last *two* sentinels we found, and clamp re-rereads at the second to last. This make us reread at most one corrupt record.